csharp/src/Ice/Reference.cs

```diff
@@ -745,149 +868,143 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
-                                           "the request size exceeds Ice.RetryRequestSizeMax, " :
-                                           "the retry buffer size would exceed Ice.RetryBufferSizeMax, ") +
-                                           "passing exception through to the application",
-                                           attempt,
-                                           retryPolicy,
-                                           lastException);
+                                excludedConnectors ??= new();
+                                excludedConnectors.Add(connectors[nextConnector]);
+                                connectors.RemoveAt(nextConnector);
                             }
-                            break; // We cannot retry, get out of the loop
-                        }
-                        else if (retryPolicy == RetryPolicy.NoRetry)
-                        {
-                            break; // We cannot retry, get out of the loop
-                        }
-                        else if (++attempt > reference.Communicator.RetryMaxAttempts)
-                        {
-                            if (reference.Communicator.TraceLevels.Retry >= 1)
+                            else if (connection?.Connector != null)
                             {
-                                TraceRetry("request failed with retryable exception but it was the final attempt,\n" +
-                                           "passing exception through to the application",
-                                           attempt,
-                                           retryPolicy,
-                                           lastException);
+                                excludedConnectors ??= new();
+                                excludedConnectors.Add(connection.Connector);
                             }
-                            break; // We cannot retry, get out of the loop
                         }
-                        else
+
+                        if (connectors != null)
                         {
-                            Debug.Assert(attempt <= reference.Communicator.RetryMaxAttempts &&
-                                         retryPolicy != RetryPolicy.NoRetry);
-                            if (retryPolicy == RetryPolicy.OtherReplica)
+                            if (connection == null && retryPolicy != RetryPolicy.OtherReplica)
+                            {
+                                // If connection establishment failed, try the next connector
+                                nextConnector = ++nextConnector % connectors.Count;
+                            }
+
+                            if (connectors.Count == 0 || nextConnector == 0)
                             {
-                                if (reference.IsFixed)
+                                // If the connector set is empty because all connectors has been excluded, or next
```

Comment:
all the connectors have been excluded and nextConnector == 0, it indicates that we already tried all the connectors

Reply:
do you mean `or nextConnector == 0, ...` instead of `and ...`

Reply:
Yes, sorry I meant "or" not "and".